contracts/tests/tests/governor/test_propose.rs

```diff
@@ -185,13 +224,127 @@ fn test_propose_snapshot() {
     assert_eq!(proposal.id, 0);
     assert_eq!(proposal.config.title, title);
     assert_eq!(proposal.config.description, description);
+    matches!(proposal.config.action, ProposalAction::Snapshot);
     assert_eq!(proposal.data.creator, samwise);
     assert_eq!(proposal.data.vote_start, e.ledger().sequence());
     assert_eq!(
         proposal.data.vote_end,
         e.ledger().sequence() + settings.vote_period
     );
-    assert_eq!(proposal.data.status, ProposalStatus::Pending);
+    assert_eq!(proposal.data.status, ProposalStatus::Open);
+}
+
+#[test]
+fn test_propose_upgrade() {
+    let e = Env::default();
+    e.mock_all_auths();
+
+    let bombadil = Address::generate(&e);
+    let samwise = Address::generate(&e);
+    let settings = default_governor_settings(&e);
+    let (governor_address, token_address, votes_address) =
+        create_governor(&e, &bombadil, &settings);
+    let token_client = MockTokenClient::new(&e, &token_address);
+    let votes_client = TokenVotesClient::new(&e, &votes_address);
+    let governor_client = GovernorContractClient::new(&e, &governor_address);
+
+    let samwise_mint_amount: i128 = 10_000_000;
+    token_client.mint(&samwise, &samwise_mint_amount);
+    votes_client.deposit_for(&samwise, &samwise_mint_amount);
+
+    let (title, description, _) = default_proposal_data(&e);
+    let bytes = BytesN::<32>::random(&e);
+    let action = ProposalAction::Upgrade(bytes);
+
+    let proposal_id = governor_client.propose(&samwise, &title, &description, &action);
+
+    let proposal = governor_client.get_proposal(&proposal_id).unwrap();
+    assert_eq!(proposal.id, 0);
+    assert_eq!(proposal.config.title, title);
+    assert_eq!(proposal.config.description, description);
+    matches!(proposal.config.action, ProposalAction::Upgrade(_));
```

Comment:
Couldn't you just `assert_eq!` against `action`?

Reply:
Same for the test below

Reply:
No, Rust `enums` don't have `PartialEq` derived automatically. Match is able to determine if it is the same enum type, tho.

We have tests the ensures the contents are applied correctly during execution, this is just a sanity check to ensure the proper proposal was proposed.